lib/select.c

```diff
@@ -87,15 +87,52 @@ int Curl_wait_ms(int timeout_ms)
 #if defined(MSDOS)
   delay(timeout_ms);
 #elif defined(USE_WINSOCK)
-  Sleep(timeout_ms);
+  /* prevent overflow, timeout_ms is typecast to ULONG/DWORD. */
+#if TIMEDIFF_T_MAX > ULONG_MAX
+  if(timeout_ms > ULONG_MAX)
+    timeout_ms = ULONG_MAX;
+#endif
+  Sleep((ULONG)timeout_ms);
 #else
 #if defined(HAVE_POLL_FINE)
-  r = poll(NULL, 0, timeout_ms);
+  /* prevent overflow, timeout_ms is typecast to int. */
+#if TIMEDIFF_T_MAX > INT_MAX
+  if(timeout_ms > INT_MAX)
+    timeout_ms = INT_MAX;
+#endif
+  r = poll(NULL, 0, (int)timeout_ms);
 #else
   {
     struct timeval pending_tv;
-    pending_tv.tv_sec = timeout_ms / 1000;
-    pending_tv.tv_usec = (timeout_ms % 1000) * 1000;
+    timediff_t tv_sec = timeout_ms / 1000;
+    timediff_t tv_usec = (timeout_ms % 1000) * 1000;
+#ifdef WIN32
+#if TIMEDIFF_T_MAX > LONG_MAX
```

Comment:
Just to reiterate we already use Sleep for Winsock so why can't we just use that for all of Windows?

Reply:
I think the historic reasoning for this code separation there other types of Windows systems, e.g. DOS. I am not sure if that is needed anymore.